beacon-chain/state/stategen/getter.go

```diff
@@ -34,6 +35,51 @@ func (s *State) StateByRoot(ctx context.Context, blockRoot [32]byte) (*state.Bea
 	return s.loadHotStateByRoot(ctx, blockRoot)
 }
 
+// StateByRootInitialSync retrieves the state from the DB for the initial syncing phase.
+// It assumes initial syncing using a block list rather than a block tree hence the returned
+// state is not copied.
+// Do not use this method for anything other than initial syncing purpose or block tree is applied.
+func (s *State) StateByRootInitialSync(ctx context.Context, blockRoot [32]byte) (*state.BeaconState, error) {
+	// Genesis case. If block root is zero hash, short circuit to use genesis state stored in DB.
+	if blockRoot == params.BeaconConfig().ZeroHash {
+		return s.beaconDB.State(ctx, blockRoot)
+	}
+
+	if s.hotStateCache.Has(blockRoot) {
+		return s.hotStateCache.GetWithoutCopy(blockRoot), nil
```

Comment:
So doing this will invalidate the state in the hot state cache, you should delete the entry for the corresponding block root in the cache. Otherwise your justified/finalized states when retrieving from the cache will be invalid , and you might end up storing invalid states for your cold state.

Reply:
Makes sense. Addressed here:
https://github.com/prysmaticlabs/prysm/pull/5778/files#diff-f407600862245d51b706766ae42ee345R91

I can't address it within `StateByRootInitialSync` because `GetHead` also uses `StateByRootInitialSync` and I wont want to invalid there

Reply:
Ok. Found the best entry to invalidate cache. Here: https://github.com/prysmaticlabs/prysm/pull/5778/files#diff-9c12fd7b577ed54bd9355696d1b35346R210